server/jetstream_api.go

```diff
@@ -1033,6 +1080,193 @@ func (s *Server) jsStreamPurgeRequest(sub *subscription, c *client, subject, rep
+	if err != nil {
+		resp.Error = jsNotFoundError(err)
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+
+	var req JSApiStreamSnapshotRequest
+	if err := json.Unmarshal(msg, &req); err != nil {
+		resp.Error = jsInvalidJSONErr
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+	if !IsValidSubject(req.DeliverSubject) {
+		resp.Error = jsBadRequestErr
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+
+	sr, err := mset.Snapshot(0, !req.NoConsumers)
+	if err != nil {
+		resp.Error = jsError(err)
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+
+	resp.NumBlks = sr.NumBlks
+	resp.BlkSize = sr.BlkSize
+	s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(resp))
+
+	// Now do the real streaming in a separate go routine.
+	go s.streamSnapshot(c, mset, sr, &req)
+}
+
+const defaultSnapshotChunkSize = 64 * 1024
+
+// streamSnapshot will stream out our snapshot to the reply subject.
+func (s *Server) streamSnapshot(c *client, mset *Stream, sr *SnapshotResult, req *JSApiStreamSnapshotRequest) {
+	chunkSize := req.ChunkSize
+	if chunkSize == 0 {
+		chunkSize = defaultSnapshotChunkSize
+	}
+	// Setup for the chunk stream.
+	acc := c.acc
+	reply := req.DeliverSubject
+	r := sr.Reader
+	defer r.Close()
+
+	// Check interest for the snapshot deliver subject.
+	inch := make(chan bool, 1)
+	acc.sl.RegisterNotification(req.DeliverSubject, inch)
+	defer acc.sl.ClearNotification(req.DeliverSubject, inch)
+	hasInterest := <-inch
+	if !hasInterest {
+		// Allow 2 seconds or so for interest to show up.
+		select {
+		case <-inch:
+		case <-time.After(2 * time.Second):
+		}
+	}
+
+	// Create our ack flow handler.
+	// This is very simple for now.
+	acks := make(chan struct{}, 1)
+	acks <- struct{}{}
+
+	ackSubj := fmt.Sprintf("_snapshot_.%s", nuid.Next())
+	ackSub, _ := mset.subscribeInternalUnlocked(ackSubj, func(_ *subscription, _ *client, _, _ string, _ []byte) {
+		acks <- struct{}{}
+	})
+	defer mset.unsubscribeUnlocked(ackSub)
+
+	// TODO(dlc) - Add in NATS-Chunked-Sequence header
+
+	// Since this is a pipe we will gather up reads and buffer internally before sending.
+	// bufio will not help here.
+	var frag [512]byte
+	fsize := len(frag)
+	if chunkSize < fsize {
+		fsize = chunkSize
+	}
+	chunk := make([]byte, 0, chunkSize)
+
+	for {
+		n, err := r.Read(frag[:fsize])
+		// Treat all errors the same for now, just break out.
+		// TODO(dlc) - when we use headers do error if not io.EOF
+		if err != nil {
+			break
+		}
+		// Check if we should send.
+		if len(chunk)+n > chunkSize {
+			// Wait on acks for flow control.
+			// Wait up to 10ms for now if none received.
+			select {
+			case <-acks:
+			case <-inch: // Lost interest
+				goto done
+			case <-time.After(10 * time.Millisecond):
+			}
+			mset.sendq <- &jsPubMsg{reply, _EMPTY_, ackSubj, chunk, nil, 0}
```

Comment:
this being the same sendq as normal message delivery should we be concerned about backups consuming all the senq buffer and starving ongoing work?

Especially outgoing acks from stream submits as those can be quite time sensitive and we don't want to recreate the issues TCP have when the there's a lot of pressure and acks come late causing many retransmissions.

Reply:
The snapshots are pretty conservative atm. We could move them odd though if it becomes a problem, will add a comment.